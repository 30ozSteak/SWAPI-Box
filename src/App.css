@font-face {
  font-family: starWar;
  src: url(./Fonts/Starjhol.ttf);
}

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.App {
  text-align: center;
  background-color: black;
}

.App-header {
  min-height: 100vh;
  font-size: calc(10px + 2vmin);
}

.App-link {
  color: #61dafb;
}

h1 {
  color: #ffd700;
  font-family: starWar;
  font-size: 5rem;
}

.marquee-block {
  position: absolute;
  transform-origin: 50% 100%;
  width: 100%;
  bottom: 10rem;
  font-size: 7.3vw;
  text-align: justify;
  transform: perspective(300px) rotateX(25deg);
  height: 50rem;
  overflow: hidden;
}

p {
  color: #ffd700;
  position: absolute;
  animation: ⭐ 120s linear infinite;
}

@keyframes ⭐ {
  0% {
    top: 100%;
  }
  100% {
    top: -300%;
  }
}
