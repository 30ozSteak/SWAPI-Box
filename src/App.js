import React, { Component } from "react";
import "./App.css";
import Marquee from "./Marquee/Marquee";
import Menu from "./Menu/Menu";
import People from "./People/People";
import fetchData from "./Fetch/Fetch";
import filmData from "./Fetch/fetchFilm";
import getPeopleData from "./Fetch/fetchPeople";
import fetchPlanets from "./Fetch/fetchPlanets";
import vehicleData from "./Fetch/fetchVehicles";
import getSpeciesData from "./Fetch/fetchSpecies";
import getResidentData from "./Fetch/fetchResidents";
import Header from "./Header/Header";
import Loading from "./Loading/Loading";
import Planets from "./Planets/Planets";
import Vehicles from "./Vehicles/Vehicles";
import Favorites from "./Favorites/Favorites";
import getHomeWorldData from "./Fetch/fetchHomeWorld"
import { Route, NavLink } from "react-router-dom";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      films: [],
      people: [],
      residents: [],
      vehicles: [],
      planets: [],
      homeWorld: [],
      species: [],
      homeState: "active-main home-main",
      favorites: [],
      error: ''
    };
  }

  updateLocalStorage = (key, data) => {
    localStorage.setItem(key, JSON.stringify(data));
  };

  showFilmCrawl = async () => {
    if (localStorage.getItem("fetchedFilm") === null) {
      const films = await fetchData("https://swapi.co/api/films/");
      return Promise.all([films]).then(
        this.setState({
          films: films
        })
      );
    } else {
      let films = JSON.parse(localStorage.getItem("fetchedFilm"));
      this.setState({ films: films });
    }
  }

  handlePeopleLink = async () => {
    if (localStorage.getItem("fetchedPeople") === null) {
      const people = await getPeopleData();
      const planets = await fetchPlanets();
      const species = await getSpeciesData();
      let peoplePromise = Promise.all([people, planets, species]).then(
        this.setState({
          people: people,
          planets: planets,
          species: species
        })
      );
      this.updateLocalStorage("fetchedPeople", this.state.people);
      this.updateLocalStorage("fetchedPlanet", this.state.planets);
      this.updateLocalStorage("fetchedSpecies", this.state.species);
      console.log('if statement', this.state.planets)
      return peoplePromise;
    } else {
      let peopleData = JSON.parse(localStorage.getItem("fetchedPeople"));
      let planetsData = JSON.parse(localStorage.getItem("fetchedPlanet"));
      let speciesData = JSON.parse(localStorage.getItem("fetchedSpecies"));
      this.setState({
        people: peopleData,
        planets: planetsData,
        species: speciesData
      });
      console.log('else statement', this.state.planets)
    }
  };

  handleVehicleLink = async () => {
    if (localStorage.getItem("fetchedVehicles") === null) {
      const vehicles = await fetchData("https://swapi.co/api/vehicles");
      let promiseVehicles = Promise.all([vehicles]).then(
        this.setState({
          vehicles: vehicles
        })
      );
      this.updateLocalStorage("fetchedVehicles", this.state.vehicles);
      return promiseVehicles;
    } else {
      let vehicleData = localStorage.getItem("fetchedVehicles");
      let vehicles = JSON.parse(vehicleData);
      this.setState({
        vehicles: vehicles
      });
    }
  };

  handlePlanetLink = async () => {
    if (localStorage.getItem("fetchedResidents") === null) {
      const homeWorld = await getHomeWorldData();
      const residents = await getResidentData();
      let promisedData = Promise.all([homeWorld, residents]).then(
        this.setState({
          homeWorld: homeWorld,
          residents: residents
        })
      );
      this.updateLocalStorage("fetchedHomeWorld", this.state.homeWorld);
      this.updateLocalStorage("fetchedResidents", this.state.residents);
      console.log('if statement', this.state.homeWorld)
      return promisedData
    } else {
      let homeWorldData = JSON.parse(localStorage.getItem('fetchedHomeWorld'))
      let residentsData = JSON.parse(localStorage.getItem('fetchedResidents'))
      this.setState({
        residents: residentsData,
        homeWorld: homeWorldData
      });
      console.log('else statement', this.state.homeWorld)
    }
  };

  async componentDidMount() {
    await this.checkPathName();
    await this.showFilmCrawl();
    if (localStorage.getItem("fetchedFilm") === null) {
      this.updateLocalStorage("fetchedFilm", this.state.films);
    } else {
      return;
    }
  }

  handleFavorites = async (favorite) => {
    const newFavorite = { ...favorite, id: favorite.name }
    await this.setState({ favorites: [newFavorite, ...this.state.favorites] });
    await this.updateLocalStorage("Favorites", this.state.favorites)
    let favoritesData = JSON.parse(localStorage.getItem('Favorites'))
    await this.setState({ favorites: favoritesData })
  };

  removeFavorites = async (id) => {
    const favorites = this.state.favorites.filter(fav => fav.id !== id)
    await this.setState({ favorites })
    await this.updateLocalStorage("Favorites", this.state.favorites)
  }

  checkPathName = async () => {
    if (window.location.pathname === "/") {
      await this.showFilmCrawl();
    } else if (window.location.pathname === "/people") {
      await this.handlePeopleLink();
    } else if (window.location.pathname === "/planets") {
      await this.handlePlanetLink();
    } else if (window.location.pathname === "/vehicles") {
      await this.handleVehicleLink();
    } else if (window.location.pathname === "/favorites") {
      let favoritesData = JSON.parse(localStorage.getItem('Favorites'))
      await this.setState({ favorites: favoritesData })
    }
  };

  render() {
    return (
      <div className="App">
        <div className="header-block">
          <Header />
        </div>
        <div className="twinkle" />
        <div className="marquee-container">
          <Route
            exact
            path="/"
            render={() => <Marquee films={this.state.films} />}
          />
          <Route
            exact
            path="/people"
            render={() => (
              <People
                removeFavorites={this.removeFavorites}
                handleFavorites={this.handleFavorites}
                handlePeopleLink={this.handlePeopleLink}
                people={this.state.people}
                planets={this.state.planets}
                species={this.state.species}
                error={this.state.error}
              />
            )}
          />
          <Route
            exact
            path="/planets"
            render={() => (
              <Planets
                handleFavorites={this.handleFavorites}
                planets={this.state.homeWorld}
                residents={this.state.residents}
                error={this.state.error}
              />
            )}
          />
          <Route
            exact
            path="/vehicles"
            render={() => <Vehicles vehicles={this.state.vehicles}
              error={this.state.error}
              handleFavorites={this.handleFavorites}
            />}
          />
          <Route
            exact
            path="/favorites"
            render={() => <Favorites
              removeFavorites={this.removeFavorites}
              favorites={this.state.favorites}
              error={this.state.error}
            />}
          />
        </div>
        <Menu
          handlePlanetLink={this.handlePlanetLink}
          handlePeopleLink={this.handlePeopleLink}
          handleVehicleLink={this.handleVehicleLink}
          favorites={this.state.favorites}
        />
      </div>
    );
  }
}

export default App;
