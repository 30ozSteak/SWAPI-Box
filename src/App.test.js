import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme'
import fetchData from './Fetch/Fetch'

describe('App', () => {

it('should match the screenshot', () => {
    const wrapper = shallow(<App />)
    expect(wrapper).toMatchSnapshot()
  })

it('should have a default state', async () => {
    const wrapper = shallow(<App />)
    const expected = { films: [], homeState: 'active-main home-main', people: [], vehicles: [], planets: [], homeWorld: [], favorites: [], error: '', residents: [], species: [] }
    expect(wrapper.state()).toEqual(expected)
  })

 it('should retreive data from API for scrolling film information', async () => {

  let expected = [{ name: 'Luke Skywalker', homeworld: 'Tatooine' }]

  window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      ok: true, json: () => Promise.resolve(expected)}))

  const URL = 'https://swapi.co/api/people';
  const result = await fetchData(URL)
  const filmData = result
  expect(filmData).toEqual(expected) 
 })


})  