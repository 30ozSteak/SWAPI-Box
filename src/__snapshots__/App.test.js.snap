// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the screenshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
        >
          <h1>
            swapi-box
          </h1>
        </header>,
        <div
          className="header-block"
        >
          <Header />
        </div>,
        <div
          className="twinkle"
        />,
        <div
          className="marquee-container"
        >
          <Marquee
            films={Array []}
          />
          <People
            people={Array []}
            planets={Array []}
            species={Array []}
          />
        </div>,
        <Menu
          data={
            Array [
              Object {
                "link": "#",
                "swLink": "People",
              },
              Object {
                "link": "#",
                "swLink": "Planets",
              },
              Object {
                "link": "#",
                "swLink": "Vehicles",
              },
              Object {
                "link": "#",
                "swLink": "Favorites",
              },
            ]
          }
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1>
            swapi-box
          </h1>,
          "className": "App-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "swapi-box",
          },
          "ref": null,
          "rendered": "swapi-box",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Header />,
          "className": "header-block",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "twinkle",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Marquee
              films={Array []}
            />,
            <People
              people={Array []}
              planets={Array []}
              species={Array []}
            />,
          ],
          "className": "marquee-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "films": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "people": Array [],
              "planets": Array [],
              "species": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "data": Array [
            Object {
              "link": "#",
              "swLink": "People",
            },
            Object {
              "link": "#",
              "swLink": "Planets",
            },
            Object {
              "link": "#",
              "swLink": "Vehicles",
            },
            Object {
              "link": "#",
              "swLink": "Favorites",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
          >
            <h1>
              swapi-box
            </h1>
          </header>,
          <div
            className="header-block"
          >
            <Header />
          </div>,
          <div
            className="twinkle"
          />,
          <div
            className="marquee-container"
          >
            <Marquee
              films={Array []}
            />
            <People
              people={Array []}
              planets={Array []}
              species={Array []}
            />
          </div>,
          <Menu
            data={
              Array [
                Object {
                  "link": "#",
                  "swLink": "People",
                },
                Object {
                  "link": "#",
                  "swLink": "Planets",
                },
                Object {
                  "link": "#",
                  "swLink": "Vehicles",
                },
                Object {
                  "link": "#",
                  "swLink": "Favorites",
                },
              ]
            }
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              swapi-box
            </h1>,
            "className": "App-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "swapi-box",
            },
            "ref": null,
            "rendered": "swapi-box",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Header />,
            "className": "header-block",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "twinkle",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Marquee
                films={Array []}
              />,
              <People
                people={Array []}
                planets={Array []}
                species={Array []}
              />,
            ],
            "className": "marquee-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "films": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "people": Array [],
                "planets": Array [],
                "species": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "data": Array [
              Object {
                "link": "#",
                "swLink": "People",
              },
              Object {
                "link": "#",
                "swLink": "Planets",
              },
              Object {
                "link": "#",
                "swLink": "Vehicles",
              },
              Object {
                "link": "#",
                "swLink": "Favorites",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
